pallets/qa-tools/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+        /// and `best_ask_price` - at least 3 steps below maximum price,
+        #[pallet::call_index(1)]
+        #[pallet::weight(<T as Config>::WeightInfo::order_book_create_and_fill_many())]
+        pub fn order_book_create_and_fill_many(
+            origin: OriginFor<T>,
+            bids_owner: T::AccountId,
+            asks_owner: T::AccountId,
+            fill_settings: Vec<(OrderBookId<T::AssetId, T::DEXId>, OrderBookFillSettings)>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+
+            let order_book_ids: Vec<_> = fill_settings.iter().map(|(id, _)| id).cloned().collect();
+            Self::create_multiple_empty_unchecked(&who, order_book_ids).map_err(|e| {
+                DispatchErrorWithPostInfo {
+                    post_info: PostDispatchInfo {
+                        actual_weight: None,
+                        pays_fee: Pays::No,
+                    },
+                    error: e,
+                }
+            })?;
+            Self::fill_multiple_empty_unchecked(bids_owner, asks_owner, fill_settings).map_err(
+                |e| DispatchErrorWithPostInfo {
+                    post_info: PostDispatchInfo {
+                        actual_weight: None,
+                        pays_fee: Pays::No,
+                    },
+                    error: e,
+                },
+            )?;
+
+            // Extrinsic is only for testing, so we return all fees
+            // for simplicity.
+            Ok(PostDispatchInfo {
+                actual_weight: None,
+                pays_fee: Pays::No,
+            })
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Does not create an order book if it already exists
+        fn create_multiple_empty_unchecked(
+            who: &T::AccountId,
+            order_book_ids: Vec<OrderBookId<T::AssetId, T::DEXId>>,
+        ) -> Result<(), DispatchError> {
+            let to_create_ids: Vec<_> = order_book_ids
+                .into_iter()
+                .filter(|id| !<order_book::OrderBooks<T>>::contains_key(id))
+                .collect();
+            for order_book_id in &to_create_ids {
+                if !trading_pair::Pallet::<T>::is_trading_pair_enabled(
+                    &order_book_id.dex_id,
+                    &order_book_id.quote.into(),
+                    &order_book_id.base.into(),
+                )? {
+                    trading_pair::Pallet::<T>::register_pair(
+                        order_book_id.dex_id,
+                        order_book_id.quote.into(),
+                        order_book_id.base.into(),
+                    )?;
+                }
+                if <T as pallet::Config>::AssetInfoProvider::total_balance(
+                    &order_book_id.base,
+                    &who,
+                )? == Balance::zero()
+                {
+                    assets::Pallet::<T>::mint_unchecked(&order_book_id.base, &who, 1)?;
```

Comment:
For what?

Reply:
Btw, your test is broken by this.

Reply:
I forgot to add that it's only for nft(non-div assets)